src/libraries/Common/src/System/Net/MultiArrayBuffer.cs

```diff
@@ -0,0 +1,431 @@
+
+        public bool IsEmpty => _length == 0;
+
+        public int Length => (int)_length;
+
+        public ref byte this[int index]
+        {
+            get
+            {
+                uint uindex = (uint)index;
+                if (uindex >= _length)
+                {
+                    throw new IndexOutOfRangeException();
+                }
+
+                uint offset = _start + uindex;
+                return ref _blocks![GetBlockIndex(offset)]![GetOffsetInBlock(offset)];
+            }
+        }
+
+        public int BlockCount => (int)(GetBlockIndex(_start + _length + (BlockSize - 1)) - GetBlockIndex(_start));
+
+        public Memory<byte> GetBlock(int blockIndex)
+        {
+            if ((uint)blockIndex >= BlockCount)
+            {
+                throw new IndexOutOfRangeException();
+            }
+
+            Debug.Assert(_length > 0, "Length should never be 0 here because BlockCount would be 0");
+            Debug.Assert(_blocks is not null);
+
+            uint startInBlock = (blockIndex == 0 ? GetOffsetInBlock(_start) : 0);
+            uint endInBlock = (blockIndex == BlockCount - 1 ? GetOffsetInBlock(_start + _length - 1) + 1 : BlockSize);
+
+            Debug.Assert(0 <= startInBlock, $"Invalid startInBlock={startInBlock}. blockIndex={blockIndex}, _blocks.Length={_blocks.Length}, _start={_start}, _length={_length}");
+            Debug.Assert(startInBlock < endInBlock, $"Invalid startInBlock={startInBlock}, endInBlock={endInBlock}. blockIndex={blockIndex}, _blocks.Length={_blocks.Length}, _start={_start}, _length={_length}");
+            Debug.Assert(endInBlock <= BlockSize, $"Invalid endInBlock={endInBlock}. blockIndex={blockIndex}, _blocks.Length={_blocks.Length}, _start={_start}, _length={_length}");
+
+            return new Memory<byte>(_blocks[GetBlockIndex(_start) + blockIndex], (int)startInBlock, (int)(endInBlock - startInBlock));
+        }
+
+        public MultiMemory Slice(int start)
+        {
+            uint ustart = (uint)start;
+            if (ustart > _length)
+            {
+                throw new IndexOutOfRangeException();
+            }
+
+            return new MultiMemory(_blocks, _start + ustart, _length - ustart);
+        }
+
+        public MultiMemory Slice(int start, int length)
+        {
+            uint ustart = (uint)start;
+            uint ulength = (uint)length;
+            if (ustart > _length || ustart + ulength > _length)
+            {
+                throw new IndexOutOfRangeException();
+            }
+
+            return new MultiMemory(_blocks, _start + ustart, ulength);
+        }
+
+        public void CopyTo(Span<byte> destination)
+        {
+            if (destination.Length < _length)
+            {
+                throw new ArgumentOutOfRangeException(nameof(destination));
+            }
+
+            for (int blockIndex = 0; blockIndex < BlockCount; blockIndex++)
```

Comment:
BlockCount looks like it does some computation.  Store it into a local rather than recomputing it for each iteration?